domain/externalcontroller/state/state.go

```diff
@@ -168,51 +166,74 @@ func (st *State) ImportExternalControllers(ctx context.Context, infos []crossmod
 
 func (st *State) updateExternalControllerTx(
 	ctx context.Context,
-	tx *sql.Tx,
+	tx *sqlair.TX,
 	ci crossmodel.ControllerInfo,
 ) error {
 	cID := ci.ControllerTag.Id()
-	q := `
+	updateControllerStmt, err := sqlair.Prepare(`
 INSERT INTO external_controller (uuid, alias, ca_cert)
-VALUES (?, ?, ?)
-  ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert`
+VALUES ($ExternalController.uuid, $ExternalController.alias, $ExternalController.ca_cert)
+  ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert
+`, ExternalController{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	if _, err := tx.ExecContext(ctx, q, cID, ci.Alias, ci.CACert); err != nil {
+	externalController := ExternalController{
+		ID:     cID,
+		Alias:  sql.NullString{String: ci.Alias, Valid: true},
+		CACert: ci.CACert,
+	}
+	if err := tx.Query(ctx, updateControllerStmt, externalController).Run(); err != nil {
 		return errors.Trace(err)
 	}
 
-	addrsBinds, addrsAnyVals := database.SliceToPlaceholder(ci.Addrs)
-	q = fmt.Sprintf(`
+	deleteStmt, err := sqlair.Prepare(`
 DELETE FROM external_controller_address
-WHERE  controller_uuid = ?
-AND    address NOT IN (%s)`[1:], addrsBinds)
-
-	args := append([]any{cID}, addrsAnyVals...)
-	if _, err := tx.ExecContext(ctx, q, args...); err != nil {
+WHERE  controller_uuid = $ExternalController.uuid
+AND    address NOT IN ($ControllerUUIDs[:])
+`, ExternalController{}, ControllerUUIDs{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cIDs := ControllerUUIDs(ci.Addrs)
+	if len(cIDs) == 0 {
```

Comment:
This `len` check can be removed once https://github.com/canonical/sqlair/pull/136 is in